waiter/src/waiter/util/utils.clj

```diff
@@ -527,6 +528,39 @@
                   (sleep delay-ms)
                   (recur (inc num-tries) (* delay-ms delay-multiplier))))))))
 
+(defn async-retry-strategy
+  "Return a async retry function using the specified retry config.
+   The returned function accepts a no-args async body-function to be executed. The returned function returns a channel
+   where the result of the retry strategy will be sent to.
+   The body-function must return a channel (i.e. async/go). If the channel receives an Exception, the body-function
+   will be considered for retry. Otherwise the message received from the channel will be returned.
+
+   `delay-multiplier` each previous delay is multiplied by delay-multiplier to generate the next delay.
+   `initial-delay-ms` the initial delay for the first retry.
+   `max-delay-ms` the delay cap for exponential backoff delay.
+   `max-retries`  limit the number of retries.
+   "
+  [{:keys [delay-multiplier initial-delay-ms max-delay-ms max-retries]
+    :or {delay-multiplier 1.0
+         initial-delay-ms 100
+         max-delay-ms 300000 ; 300k millis = 5 minutes
+         max-retries 10}}]
+  (fn [body-function]
+    (async/go-loop [num-tries 1
+                    current-delay-ms initial-delay-ms]
+      (let [{:keys [success result]}
+            (try
+              {:success true, :result (fa/<? (body-function))}
```

Comment:
I think we can read the result and check for it being an instance of a `Throwable` instead of throwing and catching the exception.